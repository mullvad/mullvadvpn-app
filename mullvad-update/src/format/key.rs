//! Key and signature types for API version response format

use std::str::FromStr;

use anyhow::{bail, Context};
use ed25519_dalek::ed25519::signature::SignerMut;
use rand::RngCore;
use serde::{Deserialize, Serialize};

/// ed25519 secret/signing key
#[derive(Debug, Clone, PartialEq)]
pub struct SecretKey(pub ed25519_dalek::SecretKey);

impl SecretKey {
    /// Generate a new secret ed25519 key
    pub fn generate() -> Self {
        // Using OsRng is suggested by the docs
        let mut bytes = ed25519_dalek::SecretKey::default();
        rand::rngs::OsRng.fill_bytes(&mut bytes);
        SecretKey(bytes)
    }

    pub fn pubkey(&self) -> VerifyingKey {
        let sign_key = ed25519_dalek::SigningKey::from_bytes(&self.0);
        VerifyingKey(sign_key.verifying_key())
    }

    /// Sign data using this key
    pub fn sign(&self, msg: &[u8]) -> Signature {
        let mut secret = ed25519_dalek::SigningKey::from_bytes(&self.0);
        Signature(secret.sign(msg))
    }
}

impl ToString for SecretKey {
    fn to_string(&self) -> String {
        hex::encode(self.0)
    }
}

impl<'de> Deserialize<'de> for SecretKey {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let key = String::deserialize(deserializer)?;
        let key = bytes_from_hex::<{ ed25519_dalek::SECRET_KEY_LENGTH }>(&key)
            .map_err(|err| serde::de::Error::custom(err.to_string()))?;
        Ok(SecretKey(key))
    }
}

impl FromStr for SecretKey {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let bytes = bytes_from_hex::<{ ed25519_dalek::SECRET_KEY_LENGTH }>(&s)?;
        Ok(SecretKey(bytes))
    }
}

impl Serialize for SecretKey {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&hex::encode(self.0))
    }
}

/// ed25519 verifying key
#[derive(Debug, PartialEq, Eq)]
pub struct VerifyingKey(pub ed25519_dalek::VerifyingKey);

impl<'de> Deserialize<'de> for VerifyingKey {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let bytes = String::deserialize(deserializer)?;
        let bytes = bytes_from_hex::<{ ed25519_dalek::PUBLIC_KEY_LENGTH }>(&bytes)
            .map_err(|err| serde::de::Error::custom(err.to_string()))?;
        let key = ed25519_dalek::VerifyingKey::from_bytes(&bytes).map_err(|_err| {
            serde::de::Error::invalid_value(
                serde::de::Unexpected::Other("invalid verifying key"),
                &"valid ed25519 key",
            )
        })?;
        Ok(VerifyingKey(key))
    }
}

impl Serialize for VerifyingKey {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&hex::encode(self.0.as_bytes()))
    }
}

/// ed25519 signature
pub struct Signature(pub ed25519_dalek::Signature);

impl<'de> Deserialize<'de> for Signature {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let bytes = String::deserialize(deserializer)?;
        let bytes = bytes_from_hex::<{ ed25519_dalek::SIGNATURE_LENGTH }>(&bytes)
            .map_err(|err| serde::de::Error::custom(err.to_string()))?;
        Ok(Signature(ed25519_dalek::Signature::from_bytes(&bytes)))
    }
}

impl Serialize for Signature {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&hex::encode(self.0.to_bytes()))
    }
}

/// Deserialize a hex-encoded string to a bytes array of an exact size
fn bytes_from_hex<const SIZE: usize>(key: &str) -> anyhow::Result<[u8; SIZE]> {
    let bytes = hex::decode(key).context("invalid hex")?;
    if bytes.len() != SIZE {
        bail!(
            "hex-encoded string of {SIZE} bytes, found {} bytes",
            bytes.len()
        );
    }
    let mut key = [0u8; SIZE];
    key.copy_from_slice(&bytes);
    Ok(key)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_serialization_and_deserialization() {
        let secret = SecretKey::generate();
        let pubkey = secret.pubkey();

        let secret_hex = hex::encode(secret.0);
        let pubkey_hex = hex::encode(pubkey.0);

        // Test serialization
        let actual = serde_json::json!({
            "secret": secret,
            "key": pubkey,
        });
        let expected: serde_json::Value = serde_json::from_str(&format!(
            r#"{{
            "secret": "{secret_hex}",
            "key": "{pubkey_hex}"
        }}"#
        ))
        .unwrap();

        assert_eq!(actual, expected);

        // Test deserialization
        let secret_obj = actual.as_object().unwrap().get("secret").unwrap().clone();
        let deserialized_secret: SecretKey = serde_json::from_value(secret_obj).unwrap();

        let pubkey_obj = actual.as_object().unwrap().get("key").unwrap().clone();
        let deserialized_pubkey: VerifyingKey = serde_json::from_value(pubkey_obj).unwrap();

        assert_eq!(deserialized_secret, secret);
        assert_eq!(deserialized_pubkey, pubkey);
    }
}
