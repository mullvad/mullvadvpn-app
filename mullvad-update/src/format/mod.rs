//! This module includes all that is needed for the (de)serialization of Mullvad version metadata.
//! This includes ensuring authenticity and integrity of version metadata, and rejecting expired
//! metadata. There are also tools for producing new versions.
//!
//! Fundamentally, a version object is a JSON object with a `signed` key and a `signature` key.
//! `signature` contains a public key and an ed25519 signature of `signed` in canonical JSON form.
//! `signed` also contains an `expires` field, which is a timestamp indicating when the object
//! expires.
//!
//! For [deserializer] to succeed in deserializing a file, it must verify that the canonicalized
//! form of `signed` is in fact signed by key/signature in `signature`. It also reads the `expires`
//! and rejects the file if it has expired.

use serde::{Deserialize, Serialize};

pub mod deserializer;
pub mod key;
#[cfg(feature = "sign")]
pub mod serializer;

/// JSON response including signature and signed content
/// This type does not implement [serde::Deserialize] to prevent accidental deserialization without
/// signature verification.
#[derive(Serialize)]
pub struct SignedResponse {
    /// Signature of the canonicalized JSON of `signed`
    pub signature: ResponseSignature,
    /// Content signed by `signature`
    pub signed: Response,
}

/// Helper class that leaves the signed data untouched
/// Note that deserializing doesn't verify anything
#[derive(Deserialize, Serialize)]
struct PartialSignedResponse {
    /// Signature of the canonicalized JSON of `signed`
    pub signature: ResponseSignature,
    /// Content signed by `signature`
    pub signed: serde_json::Value,
}

/// Signed JSON response, not including the signature
#[derive(Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Response {
    /// When the signature expires
    pub expires: chrono::DateTime<chrono::Utc>,
    /// Stable version response
    pub stable: VersionResponse,
    /// Beta version response
    pub beta: Option<VersionResponse>,
}

#[derive(Deserialize, Serialize)]
pub struct VersionResponse {
    /// The current version in this channel
    pub current: SpecificVersionResponse,
    /// The version being rolled out in this channel
    pub next: Option<NextSpecificVersionResponse>,
}

#[derive(Deserialize, Serialize)]
pub struct NextSpecificVersionResponse {
    /// The percentage of users that should receive the new version.
    pub rollout: f32,
    #[serde(flatten)]
    pub version: SpecificVersionResponse,
}

#[derive(Deserialize, Serialize)]
pub struct SpecificVersionResponse {
    /// Mullvad app version
    pub version: mullvad_version::Version,
    /// Changelog entries
    pub changelog: String,
    /// Installer details for different architectures
    pub installers: SpecificVersionArchitectureResponses,
}

/// Version details for supported architectures
#[derive(Deserialize, Serialize)]
pub struct SpecificVersionArchitectureResponses {
    /// Details for x86 installer
    pub x86: Option<SpecificVersionArchitectureResponse>,
    /// Details for ARM64 installer
    pub arm64: Option<SpecificVersionArchitectureResponse>,
}

#[derive(Deserialize, Serialize)]
pub struct SpecificVersionArchitectureResponse {
    /// Mirrors that host the artifact
    pub urls: Vec<String>,
    /// Size of the installer, in bytes
    pub size: usize,
    /// Hash of the installer, hexadecimal string
    pub sha256: String,
}

/// JSON response signature
#[derive(Deserialize, Serialize)]
pub struct ResponseSignature {
    pub keyid: key::VerifyingKey,
    pub sig: key::Signature,
}
