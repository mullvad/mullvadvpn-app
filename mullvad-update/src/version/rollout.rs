//! TODO

use std::cmp::Ordering;
use std::fmt::{self, Display};
use std::ops::RangeInclusive;
use std::str::FromStr;

use anyhow::bail;
use serde::{Deserialize, Serialize, de::Error};

/// Accept *any* version (rollout >= 0) when querying for app info.
pub const IGNORE: Rollout = Rollout(0.);

/// Accept any version (rollout > 0) when querying for app info.
/// Only versions with a non-zero rollout are supported.
pub const SUPPORTED_VERSION: Rollout = Rollout(f32::EPSILON);

/// Accept only fully rolled out versions (rollout >= 1) when querying for app info.
pub const FULLY_ROLLED_OUT: Rollout = Rollout(1.);

pub const VALID_ROLLOUT: RangeInclusive<f32> = 0.0..=1.0;

/// Rollout threshold. Any version in the response below this threshold will be ignored
///
/// INVARIANT: The inner f32 must be in the `VALID_ROLLOUT` range.
#[derive(Debug, Clone, Copy, PartialOrd, PartialEq)]
pub struct Rollout(f32);

impl Rollout {
    /// Calculate the threshold used to determine if a client is included in the current rollout of
    /// some release.
    ///
    /// Invariant: 0.0 < threshold <= 1.0
    ///
    /// 0.0 is a special-cased rollout value reserved for complete rollbacks. See [IGNORE].
    pub fn threshold(rollout_threshold_seed: u32, version: mullvad_version::Version) -> Self {
        use rand::{Rng, SeedableRng, rngs::SmallRng};
        use sha2::{Digest, Sha256};
        let mut hasher = Sha256::new();
        hasher.update(rollout_threshold_seed.to_string());
        hasher.update(version.to_string());
        let hash = hasher.finalize();
        let seed: &[u8; 32] = hash.first_chunk().expect("SHA256 hash is 32 bytes");
        let mut rng = SmallRng::from_seed(*seed);
        let threshold = rng.random_range(SUPPORTED_VERSION.0..=FULLY_ROLLED_OUT.0);
        Self::try_from(threshold).expect("threshold is within the Rollout domain")
    }
}

impl TryFrom<f32> for Rollout {
    type Error = anyhow::Error;

    fn try_from(rollout: f32) -> Result<Self, Self::Error> {
        if !rollout.is_finite() {
            bail!("rollout value must be a finite number, but was {rollout}");
        }

        if !VALID_ROLLOUT.contains(&rollout) {
            bail!(
                "rollout value {rollout} is outside valid range {}..={}",
                VALID_ROLLOUT.start(),
                VALID_ROLLOUT.end(),
            );
        }

        Ok(Rollout(rollout))
    }
}

impl Eq for Rollout {}

#[allow(clippy::derive_ord_xor_partial_ord)] // we impl Ord in terms of PartalOrd, so it's fine
impl Ord for Rollout {
    fn cmp(&self, other: &Self) -> Ordering {
        debug_assert!(self.0.is_finite());
        debug_assert!(other.0.is_finite());
        self.partial_cmp(other).expect("rollout is always in 0..=1")
    }
}

impl Display for Rollout {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl FromStr for Rollout {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let rollout: f32 = s.parse()?;
        Rollout::try_from(rollout)
    }
}

impl<'de> Deserialize<'de> for Rollout {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let rollout = f32::deserialize(deserializer)?;

        Rollout::try_from(rollout)
            .map_err(|e| e.to_string())
            .map_err(D::Error::custom)
    }
}

impl Serialize for Rollout {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.0.serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::{assert_snapshot, assert_yaml_snapshot};
    use proptest::prelude::*;

    /// Generate *any* arbitrary [Rollout] values.
    ///
    /// This generator assume that [VALID_ROLLOUT] represent all valid rollouts.
    fn arb_any_rollout() -> BoxedStrategy<Rollout> {
        VALID_ROLLOUT.prop_map(Rollout).boxed()
    }

    /// Generate an arbitrary [Rollout] values.
    ///
    /// This generator is heavily biased towards edge cases such as zero rollout, full rollout etc.
    fn arb_rollout() -> impl Strategy<Value = Rollout> {
        let any = arb_any_rollout();
        let edge_cases = prop_oneof![
            Just(IGNORE),            // = 0
            Just(SUPPORTED_VERSION), // > 0
            Just(FULLY_ROLLED_OUT)   // = 1
        ];
        // Let's say that any of the edge-cases should be generated with a 1/5 probabilty.
        prop_oneof![
            80 => any,
            20 => edge_cases
        ]
    }

    proptest! {
         /// Assert that all rollout values from 0 up to 1 are valid rollouts.
         #[test]
         fn valid_rollout(r in arb_rollout()) {
             Rollout::from_str(&r.to_string()).unwrap();
         }
    }

    const GOOD_ROLLOUT_EXAMPLES: &[f32] = &[
        -0.0,                // 0%
        0.0,                 // 0%
        -0.0 + f32::EPSILON, // > 0%
        1.0 / 3.0,           // 33%
        1.0 - f32::EPSILON,  // 99%
        1.0,                 // 100%
    ];

    const BAD_ROLLOUT_EXAMPLES: &[f32] = &[
        -f32::EPSILON,
        1.0 + f32::EPSILON,
        f32::NAN,
        f32::INFINITY,
        f32::NEG_INFINITY,
        100.0,
    ];

    #[test]
    fn test_rollout_serialization() {
        for &valid_rollout in GOOD_ROLLOUT_EXAMPLES {
            let serialized_f32 = serde_json::to_string(&valid_rollout).unwrap();
            let deserialized_rollout: Rollout = serde_json::from_str(&serialized_f32).unwrap();
            let serialized_rollout = serde_json::to_string(&deserialized_rollout).unwrap();

            assert_eq!(deserialized_rollout.0, valid_rollout);
            assert_eq!(serialized_rollout, serialized_f32);
        }
    }

    #[test]
    fn test_rollout_deserialize_bad() {
        for &bad_rollout in BAD_ROLLOUT_EXAMPLES {
            let rollout_str = bad_rollout.to_string();
            serde_json::from_str::<Rollout>(&rollout_str)
                .expect_err("must fail to deserialize bad rollout");
        }
    }

    /// Test that the `Display` impl of [Rollout] makes sense.
    /// Note clap requires that `Display` must be the inverse of `FromStr`.
    #[test]
    fn test_rollout_display() {
        let string_reprs = GOOD_ROLLOUT_EXAMPLES
            .iter()
            .map(|&f| format!("{f} => {}\n", Rollout::try_from(f).unwrap()))
            .collect::<String>();

        assert_snapshot!(&string_reprs);
    }

    #[test]
    /// Check that the implementation of [rollout_threshold] yields different threshold values as
    /// app version number progresses.
    ///
    /// Note that there is a chance for repetition - we are effectively mapping a 256 byte hash to
    /// the fractional part of an [f32], which is a much smaller domain.
    fn test_rollout_threshold_uniqueness() {
        let seed = 4; // Chosen by fair dice roll. Guaranteed to be random.
        let v20254: mullvad_version::Version = "2025.4".parse().unwrap();
        let v20255: mullvad_version::Version = "2025.5".parse().unwrap();
        assert_ne!(
            Rollout::threshold(seed, v20254.clone()),
            Rollout::threshold(seed, v20255.clone())
        );
        assert_yaml_snapshot!(Rollout::threshold(seed, v20254));
        assert_yaml_snapshot!(Rollout::threshold(seed, v20255));
    }
}
