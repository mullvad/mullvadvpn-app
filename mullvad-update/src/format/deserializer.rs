//! Deserializer and verifier of version metadata

use anyhow::Context;

use super::key::*;
use super::Response;
use super::{PartialSignedResponse, SignedResponse};

impl SignedResponse {
    /// Deserialize some bytes to JSON, and verify them, including signature and expiry.
    /// If successful, the deserialized data is returned.
    pub fn deserialize_and_verify(key: VerifyingKey, bytes: &[u8]) -> Result<Self, anyhow::Error> {
        Self::deserialize_and_verify_at_time(key, bytes, chrono::Utc::now())
    }

    /// This method is used for testing, and skips all verification.
    /// Own method to prevent accidental misuse.
    #[cfg(test)]
    pub fn deserialize_and_verify_insecure(bytes: &[u8]) -> Result<Self, anyhow::Error> {
        let partial_data: PartialSignedResponse =
            serde_json::from_slice(bytes).context("Invalid version JSON")?;
        let signed = serde_json::from_value(partial_data.signed)
            .context("Failed to deserialize response")?;
        Ok(Self {
            signature: partial_data.signature,
            signed,
        })
    }

    /// Deserialize some bytes to JSON, and verify them, including signature and expiry.
    /// If successful, the deserialized data is returned.
    fn deserialize_and_verify_at_time(
        key: VerifyingKey,
        bytes: &[u8],
        current_time: chrono::DateTime<chrono::Utc>,
    ) -> Result<Self, anyhow::Error> {
        // Deserialize and verify signature
        let partial_data = deserialize_and_verify(&key, bytes)?;

        // Deserialize the canonical JSON to structured representation
        let signed_response: Response = serde_json::from_value(partial_data.signed)
            .context("Failed to deserialize response")?;

        // Reject time if the data has expired
        if current_time >= signed_response.expires {
            anyhow::bail!(
                "Version metadata has expired: valid until {}",
                signed_response.expires
            );
        }

        Ok(SignedResponse {
            signature: partial_data.signature,
            signed: signed_response,
        })
    }
}

/// Deserialize arbitrary JSON object with a signature attached.
/// WARNING: This only verifies the signature, not expiration.
///
/// On success, this returns verified data and signature
pub(super) fn deserialize_and_verify(
    key: &VerifyingKey,
    bytes: &[u8],
) -> anyhow::Result<PartialSignedResponse> {
    let partial_data: PartialSignedResponse =
        serde_json::from_slice(bytes).context("Invalid version JSON")?;

    // Check if the key matches
    if partial_data.signature.keyid.0 != key.0 {
        anyhow::bail!("Unrecognized key");
    }

    // Serialize to canonical json format
    let canon_data = json_canon::to_vec(&partial_data.signed)
        .context("Failed to serialize to canonical JSON")?;

    // Check if the data is signed by our key
    partial_data
        .signature
        .keyid
        .0
        .verify_strict(&canon_data, &partial_data.signature.sig.0)
        .context("Signature verification failed")?;

    Ok(PartialSignedResponse {
        signature: partial_data.signature,
        // Serialize back in case something was lost during deserialization
        signed: serde_json::from_slice(&canon_data)
            .context("Failed to serialize canonical JSON")?,
    })
}

#[cfg(test)]
mod test {
    use std::str::FromStr;

    use super::*;

    /// Test that a valid signed version response is successfully deserialized and verified
    #[test]
    fn test_response_deserialization_and_verification() {
        const TEST_PUBKEY: &str =
            "f4c262705b4ae8088bc8173889f779f77563edfd7de3b0ac4aa0e554a6896404";
        let pubkey = hex::decode(TEST_PUBKEY).unwrap();
        let verifying_key =
            ed25519_dalek::VerifyingKey::from_bytes(&pubkey.try_into().unwrap()).unwrap();

        SignedResponse::deserialize_and_verify_at_time(
            VerifyingKey(verifying_key),
            include_bytes!("../../test-version-response.json"),
            // It's 1970 again
            chrono::DateTime::UNIX_EPOCH,
        )
        .expect("expected valid signed version metadata");

        // Reject expired data
        SignedResponse::deserialize_and_verify_at_time(
            VerifyingKey(verifying_key),
            include_bytes!("../../test-version-response.json"),
            // In the year 3000
            chrono::DateTime::from_str("3000-01-01T00:00:00Z").unwrap(),
        )
        .expect_err("expected expired version metadata");
    }
}
