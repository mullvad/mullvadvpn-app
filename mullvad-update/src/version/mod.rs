//! This module is used to extract the latest versions out of a raw [format::Response] using a query
//! [VersionParameters]. It also contains additional logic for filtering and validating the raw
//! deserialized response.
//!
//! The main input here is [VersionParameters], and the main output is [VersionInfo].

pub mod rollout;

use std::cmp::Ordering;

use anyhow::{Context, bail};
use itertools::Itertools;

use crate::format::{self, Installer as AppInstaller, Release};
use rollout::Rollout;

/// Lowest version to accept using 'verify'
pub const MIN_VERIFY_METADATA_VERSION: usize = 0;

/// Query type for [VersionInfo]
#[derive(Debug, Clone)]
pub struct VersionParameters {
    /// Architecture to retrieve data for
    pub architecture: VersionArchitecture,
    /// Rollout threshold. Any version in the response below this threshold will be ignored
    pub rollout: Rollout,
    /// Allow versions without any installers to be returned
    pub allow_empty: bool,
    /// Lowest allowed `metadata_version` in the version data
    /// Typically the current version plus 1
    pub lowest_metadata_version: usize,
}

/// Installer architecture
pub type VersionArchitecture = format::Architecture;

/// Version update information derived from querying a [format::Response] and filtering with [VersionParameters]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
pub struct VersionInfo {
    /// Stable version info
    pub stable: Version,
    /// Beta version info (if available and newer than `stable`).
    /// If latest stable version is newer, this will be `None`.
    pub beta: Option<Version>,
}

/// Contains information about a version for the current target
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
pub struct Version {
    /// Version
    pub version: mullvad_version::Version,
    /// App installer
    #[serde(flatten)]
    installer: Installer,
    /// Version changelog
    pub changelog: String,
}

impl Version {
    /// Create a [Version] which does not contain an installer.
    ///
    /// This is useful for broadcasting version info on Linux, where installers will never exist.
    pub fn suckless(version: mullvad_version::Version, changelog: String) -> Self {
        Version {
            version,
            changelog,
            installer: Installer::empty(),
        }
    }

    /// Returns the chronological order between two app versions.
    ///
    /// Invariant: the versions compared ought to be of the same kind, i.e. stable, beta or alpha.
    // F: FnMut(&Self::Item, &Self::Item) -> Ordering
    pub fn latest(&self, other: &Self) -> Ordering {
        self.version
            .partial_cmp(&other.version)
            .unwrap_or(Ordering::Equal)
    }

    /// True if self is a stable app version.
    pub const fn stable(&self) -> bool {
        self.version.is_stable()
    }

    /// True if self is a beta app version.
    pub const fn beta(&self) -> bool {
        self.version.is_beta()
    }
}

/// TODO: Document
#[derive(Debug, Clone, Default, PartialEq, serde::Serialize, serde::Deserialize)]
struct Installer {
    /// URLs to use for downloading the app installer
    urls: Vec<String>,
    /// Size of installer, in bytes
    size: usize,
    /// App installer checksum
    sha256: [u8; 32],
}

impl Installer {
    fn new(urls: Vec<String>, size: usize, sha256: String) -> anyhow::Result<Self> {
        Ok(Self {
            size,
            urls,
            sha256: hex::decode(sha256)
                .context("Invalid checksum hex")?
                .try_into()
                .map_err(|_| anyhow::anyhow!("Invalid checksum length"))?,
        })
    }

    fn empty() -> Self {
        Self::default()
    }
}

impl VersionInfo {
    /// Convert signed response data to public version type
    /// NOTE: `response` is assumed to be verified and untampered. It is not verified.
    //
    // TODO: Decompose
    pub fn try_from_response(
        params: &VersionParameters,
        response: format::Response,
    ) -> anyhow::Result<Self> {
        // Check this before anything else so that it's rejected independently of `params`.
        Self::validate_releases(&response.releases)?;

        let version_info = get_version_info(
            response.releases,
            params.rollout,
            params.allow_empty,
            params.architecture,
        )?;

        Ok(version_info)
    }

    /// Fail if there are duplicate versions.
    fn validate_releases(releases: &[Release]) -> anyhow::Result<()> {
        if !releases.iter().map(|r| &r.version).all_unique() {
            bail!("API response contains multiple release for the same version");
        }
        Ok(())
    }
}

/// TODO: Document this function *very well*.
/// Input: TODO
/// Output: TODO
fn get_version_info(
    releases: Vec<Release>,
    rollout: Rollout,
    allow_empty: bool,
    architecture: format::Architecture,
) -> anyhow::Result<VersionInfo> {
    let available_versions: Vec<Version> =
        get_version_info_intermediary(releases, rollout, allow_empty, architecture)?;
    let stable = get_latest_version(&available_versions, Version::stable)
        .context("No stable version found")?;
    // Find the latest beta version.
    //
    // If the latest beta version is older than latest stable, dispose of it.
    let beta = get_latest_version(&available_versions, |app| {
        app.beta() && app.version > stable.version
    });
    Ok(VersionInfo {
        stable: stable.clone(),
        beta: beta.cloned(),
    })
}

/// Filter releases based on rollout, architecture and dev versions.
///
/// Properties:
/// - If no release in `releases`
fn get_version_info_intermediary(
    releases: Vec<Release>,
    rollout: Rollout,
    allow_empty: bool,
    architecture: format::Architecture,
) -> anyhow::Result<Vec<Version>> {
    // Map a [Release] into an [AppInstaller]
    let something = |release: Release| {
        // TODO: Can this edge-case be replaced?
        if release.installers().is_empty() && allow_empty {
            return Some(Ok(Version::suckless(release.version, release.changelog)));
        }
        // Map each artifact to a [Version]
        let version = release.version.clone();
        let changelog = release.changelog.clone();
        release.into_installer(architecture).map(
            |AppInstaller {
                 urls, size, sha256, ..
             }| {
                let installer = Installer::new(urls, size, sha256)?;
                Ok(Version {
                    version,
                    changelog,
                    installer,
                })
            },
        )
    };

    // Filter releases based on rollout, architecture and dev versions.
    releases
        .into_iter()
        // Filter out releases that are not rolled out to us
        .filter(|release| release.rollout >= rollout)
        // Filter out dev versions
        .filter(|release| !release.version.is_dev())
        .flat_map(something).try_collect()
}

/// Find latest version of some kind (i.e. stable, beta, alpha).
//
// TODO: Define properties of this function.
fn get_latest_version(
    available_versions: &[Version],
    kind: impl Fn(&Version) -> bool,
) -> Option<&Version> {
    available_versions
        .iter()
        .filter(|v| kind(v))
        .max_by(|this, that| Version::latest(this, that))
}

/// A version is considered supported if the version exists in the metadata. Versions with a
/// rollout of 0 is still considered supported.
// TODO: Define properties
pub fn is_version_supported(
    current_version: mullvad_version::Version,
    releases: &[Release],
) -> bool {
    releases
        .iter()
        .any(|release| release.version.eq(&current_version))
}

/// Generate a special seed used to calculate at which rollout percentage a client should be
/// notified about a new release.
///
/// See [Rollout::threshold] for details.
pub fn generate_rollout_seed() -> u32 {
    rand::random()
}

#[cfg(test)]
pub mod arbitrary {
    use super::*;

    use format::Architecture;
    use mullvad_version::arbitrary::arb_version as arb_app_version;
    use prop::collection;
    use prop::option;
    use proptest::prelude::*;
    use rollout::arbitrary::*;

    prop_compose! {
        /// Generate arbitrary [VersionParameters].
        pub fn arb_version_parameters(allow_empty: bool)
            (rollout in arb_rollout(), architecture in arb_architecture(), lowest_metadata_version in 0..100usize)
            -> VersionParameters {
                // TODO: Check the `lowest_metadata_version` generations.
            VersionParameters { architecture, rollout, allow_empty, lowest_metadata_version }
        }
    }

    prop_compose! {
        /// Generate arbitrary [VersionInfo] (API responses).
        pub fn arb_version_info()
            (stable in arb_version(), beta in option::of(arb_version()))
            -> VersionInfo {
            VersionInfo { stable, beta }
        }
    }

    prop_compose! {
        /// Generate arbitrary [Version].
        pub fn arb_version()
                          (app_version in arb_app_version(), changelog: String, installer in arb_installer())
                          -> Version {
            Version {
                version: app_version,
                installer,
                changelog,
            }
        }
    }

    /// Generate an arbitrary [Installer].
    // TODO: This sucks.
    fn arb_installer() -> impl Strategy<Value = Installer> {
        // TODO
        collection::vec("XD".boxed(), 5).prop_map(|urls| {
            let size = 256; // TODO
            let sha256 = [0; 32]; // TODO
            Installer { urls, size, sha256 }
        })
    }

    /// Generate a random [Architecture] uniformly.
    fn arb_architecture() -> impl Strategy<Value = Architecture> {
        prop_oneof![Just(Architecture::X86), Just(Architecture::Arm64)]
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use std::str::FromStr;

    use insta::assert_yaml_snapshot;
    use proptest::prelude::*;

    const TEST_RESPONSE: &[u8] = include_bytes!("../../test-version-response.json");

    proptest! {
        // TODO
    }

    // These tests rely on `insta` for snapshot testing. If they fail due to snapshot assertions,
    // then most likely the snapshots need to be updated. The most convenient way to review
    // changes to, and update, snapshots is by running `cargo insta review`.

    /// Test version info response handler (rollout 1, x86)
    #[test]
    fn test_version_info_parser_x86() -> anyhow::Result<()> {
        let response = format::SignedResponse::deserialize_insecure(TEST_RESPONSE)?;

        let params = VersionParameters {
            architecture: VersionArchitecture::X86,
            rollout: rollout::FULLY_ROLLED_OUT,
            allow_empty: false,
            lowest_metadata_version: 0,
        };

        // Expect: The available latest versions for X86, where the rollout is 1.
        let info = VersionInfo::try_from_response(&params, response.signed.clone())?;

        assert_yaml_snapshot!(info);

        Ok(())
    }

    /// Test version info response handler (rollout 0.01, arm64)
    #[test]
    fn test_version_info_parser_arm64() -> anyhow::Result<()> {
        let response = format::SignedResponse::deserialize_insecure(TEST_RESPONSE)?;

        let params = VersionParameters {
            architecture: VersionArchitecture::Arm64,
            rollout: rollout::SUPPORTED_VERSION,
            allow_empty: false,
            lowest_metadata_version: 0,
        };

        let info = VersionInfo::try_from_response(&params, response.signed)?;

        // Expect: The available latest versions for arm64, where the rollout is .01.
        assert_yaml_snapshot!(info);

        Ok(())
    }

    /// Versions without installers should be returned if `allow_empty` is set
    #[test]
    fn test_version_info_empty() -> anyhow::Result<()> {
        let response = format::SignedResponse::deserialize_insecure(TEST_RESPONSE)?;

        let params = VersionParameters {
            architecture: VersionArchitecture::X86,
            rollout: rollout::SUPPORTED_VERSION,
            allow_empty: true,
            lowest_metadata_version: 0,
        };

        let info = VersionInfo::try_from_response(&params, response.signed)?;

        // Expect: The available latest versions for x86, where the rollout is .01.
        assert_yaml_snapshot!(info);

        Ok(())
    }

    /// Test whether [SUPPORTED_VERSION] ignores unsupported versions (where rollout = 0.0)
    #[test]
    fn test_version_unsupported_filtering() -> anyhow::Result<()> {
        let response = format::SignedResponse::deserialize_insecure(TEST_RESPONSE)?;

        let params = VersionParameters {
            architecture: VersionArchitecture::X86,
            rollout: rollout::SUPPORTED_VERSION,
            allow_empty: true,
            lowest_metadata_version: 0,
        };

        let info = VersionInfo::try_from_response(&params, response.signed.clone())?;

        // Expect: The available latest versions for x86, where the rollout is non-zero.
        assert_yaml_snapshot!(info);

        let params = VersionParameters {
            architecture: VersionArchitecture::X86,
            rollout: rollout::IGNORE,
            allow_empty: true,
            lowest_metadata_version: 0,
        };

        let info = VersionInfo::try_from_response(&params, response.signed)?;

        // Expect: There is an even higher version where the rollout is zero.
        assert_yaml_snapshot!(info);

        Ok(())
    }

    #[test]
    fn test_is_version_supported() -> anyhow::Result<()> {
        let response = format::SignedResponse::deserialize_insecure(TEST_RESPONSE)?;

        let supported_version = mullvad_version::Version::from_str("2025.3").unwrap();
        let supported_rollout_zero_version = mullvad_version::Version::from_str("2030.3").unwrap();
        let non_supported_version = mullvad_version::Version::from_str("2025.5").unwrap();

        assert!(is_version_supported(
            supported_version,
            &response.signed.releases
        ));
        assert!(is_version_supported(
            supported_rollout_zero_version,
            &response.signed.releases
        ));
        assert!(!is_version_supported(
            non_supported_version,
            &response.signed.releases
        ));

        Ok(())
    }
}
