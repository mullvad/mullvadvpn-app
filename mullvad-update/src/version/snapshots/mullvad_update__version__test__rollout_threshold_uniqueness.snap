---
source: mullvad-update/src/version/mod.rs
expression: "Rollout::threshold(seed, v20254)"
---
0.98061204
